File: ios/Zuko/Zuko/Net/Wire.swift
import Foundation

/// Wire protocol shared with the host daemon (v0.6+). Each message is:
///
///   [type: u8][len: u16 big-endian][payload: `len` bytes]
///
/// - 0x00 DATA    — raw terminal bytes (keystrokes up, PTY output down)
/// - 0x01 RESIZE  — payload `[cols: u16 BE][rows: u16 BE]` (client -> host,
///                  also the first frame — acts as the v0.6 handshake)
/// - 0x04 PING    — `[nonce: u64 BE]`, bidirectional (legacy; kept for
///                  backward compat with older peers, ignored by v0.6+)
/// - 0x05 PONG    — `[nonce: u64 BE]`, bidirectional (legacy; same)
///
/// The client opens the stream with a single RESIZE carrying its initial
/// terminal size. The host spawns a fresh PTY at that size and starts
/// streaming. No HELLO/WELCOME exchange, no session ids — each connection
/// gets a new PTY, killed when the connection ends.
enum Wire {
    static let data: UInt8 = 0x00
    static let resize: UInt8 = 0x01
    // 0x02 (HELLO) and 0x03 (WELCOME) were used by v0.4–v0.5 for the
    // session-resume handshake. Removed in v0.6.
    static let ping: UInt8 = 0x04
    static let pong: UInt8 = 0x05

    struct Frame {
        let type: UInt8
        let payload: [UInt8]
    }

    /// Encode one length-prefixed frame ready to write to the stream.
    static func encode(type: UInt8, payload: Data) -> Data {
        var out = Data(capacity: 3 + payload.count)
        out.append(type)
        let len = UInt16(payload.count)
        out.append(UInt8((len >> 8) & 0xFF))
        out.append(UInt8(len & 0xFF))
        out.append(payload)
        return out
    }

    static func encodeResize(cols: UInt16, rows: UInt16) -> Data {
        var payload = Data(capacity: 4)
        payload.append(UInt8((cols >> 8) & 0xFF))
        payload.append(UInt8(cols & 0xFF))
        payload.append(UInt8((rows >> 8) & 0xFF))
        payload.append(UInt8(rows & 0xFF))
        return encode(type: resize, payload: payload)
    }

    /// Try to pull one complete frame off the front of `buffer`, draining it.
    /// Returns nil when there aren't enough bytes yet.
    static func parse(_ buffer: inout Data) -> Frame? {
        guard buffer.count >= 3 else { return nil }

        // `Data.removeFirst` does not guarantee that future valid indices start
        // at integer 0. TestFlight crash CDCE664B showed `buffer[0]` trapping
        // here after at least one frame had been drained. Always walk from
        // `startIndex` instead of using absolute integer subscripts.
        let typeIndex = buffer.startIndex
        let lenHighIndex = buffer.index(after: typeIndex)
        let lenLowIndex = buffer.index(after: lenHighIndex)

        let type = buffer[typeIndex]
        let len = (Int(buffer[lenHighIndex]) << 8) | Int(buffer[lenLowIndex])
        guard buffer.count >= 3 + len else { return nil }
        let payloadStart = buffer.index(after: lenLowIndex)
        let payloadEnd = buffer.index(payloadStart, offsetBy: len)
        let payload = Array(buffer[payloadStart..<payloadEnd])
        buffer.removeSubrange(typeIndex..<payloadEnd)
        return Frame(type: type, payload: payload)
    }
}

File: zuko/src/wire.rs
//! Wire protocol shared by the host and client (and the iOS app).
//!
//! One bidirectional Iroh stream, ALPN [`ALPN`] = `zuko/1`. Every message is
//! length-prefixed so resize and data stay ordered and nothing leaks into the
//! terminal as in-band escape sequences:
//!
//! ```text
//! [type: u8][len: u16 big-endian][payload: `len` bytes]
//!   0x00 DATA   payload = raw terminal bytes (keystrokes up, PTY output down)
//!   0x01 RESIZE payload = [cols: u16 BE][rows: u16 BE]   (client -> host)
//! ```

/// ALPN used for every zuko stream.
pub const ALPN: &[u8] = b"zuko/1";

pub const TYPE_DATA: u8 = 0x00;
pub const TYPE_RESIZE: u8 = 0x01;

pub struct ParsedFrame {
    pub typ: u8,
    pub payload: Vec<u8>,
}

/// Encode one length-prefixed frame ready to write to the stream.
pub fn frame(typ: u8, payload: &[u8]) -> Vec<u8> {
    let mut f = Vec::with_capacity(3 + payload.len());
    f.push(typ);
    f.extend_from_slice(&(payload.len() as u16).to_be_bytes());
    f.extend_from_slice(payload);
    f
}

/// A `0x00 DATA` frame carrying raw terminal bytes.
pub fn data_frame(bytes: &[u8]) -> Vec<u8> {
    frame(TYPE_DATA, bytes)
}

/// A `0x01 RESIZE` frame (`client -> host`).
pub fn resize_frame(cols: u16, rows: u16) -> Vec<u8> {
    let payload = [cols.to_be_bytes(), rows.to_be_bytes()].concat();
    frame(TYPE_RESIZE, &payload)
}

/// Pull one complete length-prefixed frame off the front of `buf`, draining it.
/// Returns `None` when there aren't enough bytes yet, leaving `buf` intact.
pub fn try_parse_frame(buf: &mut Vec<u8>) -> Option<ParsedFrame> {
    if buf.len() < 3 {
        return None;
    }
    let typ = buf[0];
    let len = u16::from_be_bytes([buf[1], buf[2]]) as usize;
    if buf.len() < 3 + len {
        return None;
    }
    let payload = buf[3..3 + len].to_vec();
    buf.drain(..3 + len);
    Some(ParsedFrame { typ, payload })
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn round_trips_data_resize_and_partial() {
        let mut buf = Vec::new();
        buf.extend(data_frame(b"hi"));
        buf.extend(resize_frame(120, 40));
        // A deliberately truncated third frame (header claims 5 bytes, only 2 present).
        buf.extend_from_slice(&[0x00, 0x00, 0x05, b'x', b'y']);

        let f1 = try_parse_frame(&mut buf).unwrap();
        assert_eq!(f1.typ, TYPE_DATA);
        assert_eq!(f1.payload, b"hi");

        let f2 = try_parse_frame(&mut buf).unwrap();
        assert_eq!(f2.typ, TYPE_RESIZE);
        assert_eq!(f2.payload.len(), 4);
        let cols = u16::from_be_bytes([f2.payload[0], f2.payload[1]]);
        let rows = u16::from_be_bytes([f2.payload[2], f2.payload[3]]);
        assert_eq!((cols, rows), (120, 40));

        // Partial frame: parser must wait for more bytes, leaving the remainder intact.
        assert!(try_parse_frame(&mut buf).is_none());
        assert_eq!(buf, vec![0x00, 0x00, 0x05, b'x', b'y']);
    }

    #[test]
    fn ignores_unknown_frame_type() {
        let mut buf = vec![0x42, 0x00, 0x01, 0xFF];
        let f = try_parse_frame(&mut buf).unwrap();
        assert_eq!(f.typ, 0x42);
        assert_eq!(f.payload, vec![0xFF]);
    }
}
